### ISSUE TITLE
Order of "use_backend" lines in haproxy.cfg is inconsistent (changes every 5 seconds)

### ISSUE THREAD

**Issue by timosluis** [haproxytech/kubernetes-ingress#567]
Hello,

We use the "haproxy.org/route-acl" option for selecting backends. The haproxy controller turns this into "use_backend" rules in the haproxy.cfg file. However, it seems that the order which these lines are generated is inconsistent. And this probably also causes haproxy to reload every 5 seconds (the default sync period).

It would be nice if the order was consistent.

Thanks,

Timo

**Comment by fabianonunes** [comment]
Pass `--log=debug` to the controller to see exactly which property comparison is causing a hard reload.

**Comment by EmielBruijntjes** [comment]
```
2023/09/02 06:39:00 DEBUG   route/route.go:131 Custom Route to backend 'websockets' added, reload required
2023/09/02 06:39:00 DEBUG   route/route.go:131 Custom Route to backend 'websockets' added, reload required
2023/09/02 06:39:00 DEBUG   route/route.go:131 Custom Route to backend 'api' added, reload required
2023/09/02 06:39:00 DEBUG   route/route.go:131 Custom Route to backend 'api' added, reload required
```
And more lines like that, every 5 seconds. But there are no changes in the ingress specifications, but in the haproxy.cfg file the "use_backend" lines are placed in a different order.

**Comment by fabianonunes** [comment]
I was able to reproduce that with each synchronization operation the order of the `use_backend` is changed. However, this change in order was not enough to trigger a reload in HAProxy.

Do you mind sharing your use_backend directives?

**Comment by EmielBruijntjes** [comment]
I am not sure if I expose too much confidential info if I show the entire configuration here. Can I send it in private too?

Anyway, what *could* be a factor is that we're using a trick to make "AND" conditions to route websocket-traffic to websocket-servers. We use the following annotation to detect websocket-requests (we put "} {" in the middle to turn it into an "AND" operator):
```
annotations:
  haproxy.org/route-acl: req.hdr(Upgrade) -i -m str websocket } { req.hdr(Connection) -i -m str upgrade
```
We also have some services that filter traffic based on url parameters, but I suppose that we're not doing anything exciting here:
```
annotations:
  haproxy.org/route-acl: url_param(myparam) myvalue
```
And
```
annotations:
  haproxy.org/route-acl: url_param(otherparam) -m found
```
In total we have about 20 services for which we use such acl annotations, and it leads to a haproxy configuration like:
```
use_backend mybackend1 if { var(txn.host) -m str mydomain1.example.com }  { path -m beg / }  { req.hdr(Upgrade) -i -m str websocket } { req.hdr(Connection) -i -m str upgrade } 
use_backend mybackend2 if { var(txn.host) -m str mydomain2.example.com }  { path -m beg / }  { url_param(myparam) myvalue } 
( et cetera )
```
Seems pretty normal. But every five seconds the order of the use_backend lines in /etc/haproxy/haproxy.cfg changes, and a haproxy reload is triggered.

**Comment by EmielBruijntjes** [comment]
What could also be relevant: we have *multiple* ingresses (multiple hostnames) that all point to the same backend servers. I am looking at the code now, especially here: 

https://github.com/haproxytech/kubernetes-ingress/blob/6d188841f953e87a097e51c60542ae0b7e7112a2/pkg/route/route.go#L128

The assumption here seems to be that one backend can have only one acl. But I thinl that this is incorrect, one backend can have multiple acl's.

**Comment by fabianonunes** [comment]
I've checked this scenario where multiple Ingresses point to the same server. In this case, the `use_backend` directives are duplicated, but the check `acl != routeCond` is still false and the reload isn't triggered.

**Comment by fabianonunes** [comment]
Ah! I was able to reproduce the error. Let me see if I can find a solution.

**Comment by fabianonunes** [comment]
I've submitted PR #569. If you'd like to try it out, you can build the image with `make build`

**Comment by EmielBruijntjes** [comment]
This will solve this issue for us, which is nice.

However, I still wonder what happens if you now have two ingresses with the same hostname, but with different path or path-prefixes? I think the path should also be added to the hostID. Plus I wonder if a _map_ is the right type here, because there might be more things besides the hostname, backend and path that ingresses can have in common. Woud a _set_ not be more effective?

**Comment by fabianonunes** [comment]
> I still wonder what happens if you now have two ingresses with the same hostname, but with different path or path-prefixes

Good point. We'd have the same bug.

**Comment by fabianonunes** [comment]
Instead of using Ingress attributes (host, path, backendName, match type, etc...), I used an identifier from the Ingress itself (`$namespace-$name`) as part of the key of `CustomRoutes` map. It's on the same PR.

What do you think?

**Comment by EmielBruijntjes** [comment]
I am not sure if this is correct. Did you consider the scenario in which an ingress is changed:

- The namespace+name of the ingress stay the same.
- The ACL annotation also stays the same.
- But you do change the hostname and/or the path-prefix of the ingress.
- This *should* trigger a reload, but, if I understand the pull request correctly, it doesn't.

I would say, the customRouteID should be: "backend-hostname-path-acl", and you should not maintain a map with all the custom-routes, but a set.

**Comment by fabianonunes** [comment]
But the path, domain and any other info of the route is in the route itself. For example:

```
use_backend example_dev_http if { var(txn.host) -m str dev.example.com }  { path -m beg / }  { req.hdr(Upgrade) -i -m str websocket } { req.hdr(Connection) -i -m str upgrade }
```

In this case, if we change the domain to `dev2.example.com`, or change the path to `/abc`, the next comparison will detect this change and HAProxy will be reloaded.

If I understand correctly, you're thinking that reloading is only triggered when this customRouteID is changed. But the equality comparison happens with every cache resync and every K8S change listened to by the controller.

**Comment by fabianonunes** [comment]
The idea of the unique `customRouteID` is to avoid **route id** conflicts when we have more than one Ingress pointing to the same Service. This `customRouteID` does not influence the synchronization algorithm or the decision as to whether HAProxy will be restarted or not.

**Comment by EmielBruijntjes** [comment]
I think you are right. Does it also cover this scenario (in namespace NS):

- Ingress A points to acl-annotated service X
- Ingress A is then updated to point to service Y instead

The key in the CustomRoutes map is first "NS_A_NS_X_http" (right?), and by changing the ingress, is that key then removed from the map? And is haproxy then reloaded?

**Comment by fabianonunes** [comment]
> Does it also cover this scenario (in namespace NS):
> 
> * Ingress A points to acl-annotated service X
> * Ingress A is then updated to point to service Y instead

Yes, it covers! 

> The key in the CustomRoutes map is first "NS_A_NS_X_http" (right?), and by changing the ingress, is that key then removed from the map? And is haproxy then reloaded?

Not quite. The key won't be removed (I can change that), but the backend will. So, effectively, since the backend has been removed, this route won't be in the final haproxy.cfg.

**Comment by EmielBruijntjes** [comment]
And what if the backend is not removed because it is stlll referenced to by other ingresses, for example:
- Ingress A points to acl-annotated service X
- Ingress B points to acl-annotated service X
- Ingress C points to service Y
- Ingress A is then updated to point to service Y instead.

In that case the key is not still not removed from the CustomRoutes map, but this time there are also no backends that are being removed. Does that trigger a reload?

**Comment by fabianonunes** [comment]
Yeah, you've found another bug.

**Comment by fabianonunes** [comment]
And I think my approach won't work if we have, on the same Ingress, two paths pointing to the same Service.

Maybe will better if we store all the resulting custom routes and compare them between syncs.

We're back to square one

**Comment by fabianonunes** [comment]
I've implemented this approach in PR #569, but I'm unsure if it aligns with the project's guidelines and style. Perhaps a maintainer can provide further guidance later.

**Comment by ivanmatmati** [comment]
HI @fabianonunes , I'll check that. Thanks.
